java/src/main/java/org/signal/libsignal/metadata/signedaddress/SignedAddress.java

```diff
@@ -0,0 +1,111 @@
+ * confirming that its authenticity. The userId is also the public key with which the signature was
+ * generated, so the address can be authenticated without needing an external authority.
+ */
+public class SignedAddress {
+
+  private final byte[] serialized;
+  private final byte[] signature;
+  private final byte[] userId; // 64 bytes
+  private final int deviceId;
+
+  /**
+   * Constructs a new SignedAddress from its serialized representation. It verifies the
+   * address's signature against the contained userId + deviceId.
+   *
+   * @param serialized
+   * @throws InvalidAddressException if there was a problem deserializing the address
+   * @throws InvalidKeyException if the address's signature doesn't match what's expected based on using the userId as the public key
+   */
+  public SignedAddress(byte[] serialized) throws InvalidAddressException, InvalidKeyException {
+    try {
+      SignalProtos.SignedAddress wrapper = SignalProtos.SignedAddress.parseFrom(serialized);
+
+      if (!wrapper.hasSignature() || !wrapper.hasAddress()) {
+        throw new InvalidAddressException("Missing fields");
+      }
+
+      SignalProtos.SignedAddress.Address address = SignalProtos.SignedAddress.Address.parseFrom(wrapper.getAddress());
+
+      if (!address.hasUserId() || !address.hasDeviceId()) {
+        throw new InvalidAddressException("Missing fields");
+      }
+
+      this.userId   = address.getUserId().toByteArray();
+      this.deviceId = address.getDeviceId();
+
+      this.serialized  = serialized;
+      this.signature   = wrapper.getSignature().toByteArray();
+
+      // Using the userId as the public key, verify that the signature matches the content (deviceId and userId).
+      ECPublicKey publicKey = UserId.keyFrom(this.userId);
+      if (!Curve.verifySignature(publicKey, address.toByteArray(), this.signature)) {
+        throw new InvalidAddressException("signature verification failed");
+      };
+    } catch (InvalidProtocolBufferException | InvalidKeyException e) {
+      throw new InvalidAddressException(e);
+    }
+  }
+
+  /**
+   * Constructs a new SignedAddress using the component userId, deviceId and signing it with
+   * the provided signingKey.
+   *
+   * @param userId the user's unique ID (also their public key)
+   * @param deviceId the user's device ID (actually a uint32)
+   * @param signingKey the private key corresponding to the user's ID, used to sign the address
+   * @throws InvalidKeyException
+   */
+  public SignedAddress(byte[] userId, int deviceId, ECPrivateKey signingKey) throws InvalidKeyException {
+    byte[] address = SignalProtos.SignedAddress.Address.newBuilder()
+            .setUserId(ByteString.copyFrom(userId))
+            .setDeviceId(deviceId)
+            .build().toByteArray();
+    this.signature = Curve.calculateSignature(signingKey, address);
+    this.serialized = SignalProtos.SignedAddress.newBuilder()
```

Comment:
This is where the sender self-signs the address. The encryption still happens the way it used to, [here](https://github.com/getlantern/libsignal-metadata-java/pull/1/files#diff-96b06130bab42d5433b10b22a7b30c829cdfa722d9b28f6954d9bd7bc25cd467R78).